### PR TITLE
feat(sequelize): initial adapter

### PR DESCRIPTION
**Pull request by EXIT-ALAMERY**
Added the adapter for Sequelize ORM.

### CODE REVIEW

**Review by balazsorban44 — CHANGES_REQUESTED** [review]
Thank you so much for submitting this! I had some comments, but all in all it looks great so far!

PS.: If you are into TypeScript (totally fine if you are not interested), we now have an `Adapter` interface, that could help you type the whole adapter, by applying the interface on the Adapter function.

https://github.com/nextauthjs/next-auth/blob/0cf1823e70ce48024362670d19db09fc0bbb9727/types/adapters.d.ts#L74-L125

File: packages/sequelize/src/adapter/index.js [hunk 1]
```diff
@@ -0,0 +1,416 @@
+import "core-js/stable";
+import "regenerator-runtime/runtime";
+import { createHash, randomBytes } from "crypto";
+
+import { CreateUserError } from "./lib/errors";
```
Comment (balazsorban44) [inline packages/sequelize/src/adapter/index.js:5]:
```suggestion
import { CreateUserError } from "next-auth/errors";
```

These are included in the base library now. check out the content of the errors module, there are more type of errors as well. You can delete the `./lib/errors` file now

File: packages/sequelize/src/adapter/index.js [hunk 2]
```diff
@@ -0,0 +1,416 @@
...
+import logger from "./lib/logger";
```
Comment (balazsorban44) [inline packages/sequelize/src/adapter/index.js:6]:
```suggestion
```

You will find this in `appOptions` in `getAdapter`. You can remove `./lib/logger`

File: packages/sequelize/src/adapter/index.js [hunk 3]
```diff
@@ -0,0 +1,416 @@
...
+  const { models } = config;
+
+  function getCompoundId(providerId, providerAccountId) {
+    return createHash("sha256")
+      .update(`${providerId}:${providerAccountId}`)
+      .digest("hex");
+  }
+
+  async function getAdapter(appOptions) {
+    function debug(debugCode, ...args) {
```
Comment (balazsorban44) [inline packages/sequelize/src/adapter/index.js:18]:
```suggestion
    const logger = appOptions.logger
    function debug(debugCode, ...args) {
```

**Review by balazsorban44 — CHANGES_REQUESTED** [review]

File: packages/sequelize/example/db/models/index.js [hunk 4]
```diff
@@ -0,0 +1,35 @@
+'use strict';
+
+import Sequelize, { Model } from 'sequelize';
+import { 
+  userModel, accountModel, sessionModel, verificationRequestModel 
+} from 'next-auth-sequelize-adapter/models';
```
Comment (luke-j) [inline packages/sequelize/example/db/models/index.js:6]:
```suggestion
} from '@next-auth/sequelize-adapter';
```

File: packages/sequelize/example/db/models/index.js [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+const path = require('path');
+const basename = path.basename(__filename);
```
Comment (luke-j) [inline packages/sequelize/example/db/models/index.js:9]: Not used

```suggestion
```

File: packages/sequelize/example/db/models/user_extended.js [hunk 6]
```diff
@@ -0,0 +1,34 @@
+"use strict";
+
+// import the schema for the model you want to extend
+import { userSchema } from "next-auth-sequelize-adapter";
```
Comment (luke-j) [inline packages/sequelize/example/db/models/user_extended.js:4]:
```suggestion
import { userSchema } from "@next-auth/sequelize-adapter";
```

File: packages/sequelize/example/pages/api/auth/[...nextauth].js [hunk 7]
```diff
@@ -0,0 +1,24 @@
+import NextAuth from "next-auth";
+import Providers from "next-auth/providers";
+
+// import the local db object and the SequelizeAdapter
+import db from "../../../db/models";
+import { SequelizeAdapter } from "next-auth-sequelize-adapter";
```
Comment (luke-j) [inline packages/sequelize/example/pages/api/auth/[...nextauth].js:6]:
```suggestion
import { SequelizeAdapter } from "@next-auth/sequelize-adapter";
```

File: packages/sequelize/package.json [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+  "license": "MIT",
+  "homepage": "https://github.com/nextauthjs/adapters/tree/canary/packages/sequelize",
+  "dependencies": {
+  },
+  "devDependencies": {
+    "@babel/cli": "^7.12.10",
+    "@babel/core": "^7.12.10",
+    "@babel/preset-env": "^7.12.11"
+  }
+}
```
Comment (luke-j) [inline packages/sequelize/package.json:28]: This should have a peer dependency on sequelize, right?

**Review by luke-j — COMMENTED** [review]
Nice work, would be great to get this in.

**Comment by ndom91** [comment]
@EXIT-ALAMERY we now have official [TS types](https://github.com/nextauthjs/next-auth/blob/main/types/adapters.d.ts) for the adapters as well as a test suite all adapters should pass. Maybe these can help you finalize this adapter.

If you're still interested in getting this sequelize adapter in, I'd suggest checking out the suggestions made above and then maybe checking out the new [`basic-tests.ts`](https://github.com/nextauthjs/adapters/blob/canary/basic-tests.ts) file and get your seqeulize adapter passing :+1: 

Looking forward to this!

**Comment by balazsorban44** [comment]
@ndom91 I have talked to @EXIT-ALAMERY and unfortunately he doesn't have the opportunity to finish this up, so I am going to take this PR once I have the time.

**Comment by ndom91** [comment]
Hey guys this is looking awesome! If you can add the basic tests to this package i'd be comfortable merging and releasing a canary version!

See:
- Basic Test: https://github.com/nextauthjs/adapters/blob/canary/basic-tests.ts
- Prisma Example: https://github.com/nextauthjs/adapters/blob/canary/packages/prisma/tests/index.test.ts

**Comment by EXIT-ALAMERY** [comment]
The tests for PostgreSQL passed successfully. I'll probably write tests for MySQL or MSSQL tomorrow.

test

**Comment by ndom91** [comment]
Awesome thanks a lot!

**Comment by ndom91** [comment]
Tests for MySQL and Postgresql are in. Thanks guys! Is this alright to merge from your perspective?

Btw you still need to add sequelize to the adapter matrix in `.github/workflows/canary.yml`. 

Thanks!

**Comment by ndom91** [comment]
Hey guys I was taking a look at the tests, they don't seem to pass in CI, but that makes sense since yuo don't have any script or any way to setup the db in a CI environment. The `yarn test` script just executes `jest`. I assume in your local environment you manully created the docker container with the mysql db, seeded it, etc. right?

We need to come up with a shell script to automate that whole process in CI for both mysql and postgres. I'll try to push something here shortly :+1:

**Comment by ndom91** [comment]
So I cleaned up some of the default usernames / etc to make them match to what we usually use. Next, I added some npm scripts for each mysql and postgres to execute a bash script to setup the respective containers and execute jest tests on them once they're up.

Also none of the db drivers were listed as peerDeps or anything like that, so I added them as optional peerDeps for mysql and postgres.

Both test (mysql + postgres) pass now locally for me, starting from 0, building the containers, running migrations, running tests, etc. However on both tests I still have a ton of console.log output from what looks like the sequelize adapter itself, but I couldn't find any `console.log`'s anywhere. Do you have any idea where they're coming from? Haha

What do you think?

**Comment by EXIT-ALAMERY** [comment]
The SQL queries you see in the console get generated by the sequelize library itself so, you can disable them by setting `logging: false` in the config file.

Also, I've seen the [sequelize-typescript](https://www.npmjs.com/package/sequelize-typescript) module but, the sequelize team doesn't officially maintain it. So, I decided not to implement it but, if you believe that proceeding with it would suffice the needs of this project, then I'll take a thorough look.

Thanks!

**Comment by ndom91** [comment]
> The SQL queries you see in the console get generated by the sequelize library itself so, you can disable them by setting `logging: false` in the config file.
> 
> Also, I've seen the [sequelize-typescript](https://www.npmjs.com/package/sequelize-typescript) module but, the sequelize team doesn't officially maintain it. So, I decided not to implement it but, if you believe that proceeding with it would suffice the needs of this project, then I'll take a thorough look.
> 
> Thanks!

I only mention it because the tests seem to be failing due to missing types. I did some testing with manually declare a dummy sequelize module, but that didn't seem to help.. 

@lluia any idea whats missing here? Checkout the latest failing test.

**Comment by EXIT-ALAMERY** [comment]
We can try adding sequelize as a dependency and see if it works because I created full-fledged sequelize models specifically for the tests, so they're trying to import sequelize from the node modules, which doesn't exist.

**Comment by EXIT-ALAMERY** [comment]
> Hey @EXIT-ALAMERY, we've done a major rewrite of the adapter API to simplify things significantly. Are you still interested in getting sequelize over the finish line?

That sounds great and, I wish all the best to nextauthjs, but, unfortunately, I am no longer interested in the sequelize adapter. Thank you for your patience and support.

**Comment by balazsorban44** [comment]
no problem! we might reuse your code when we get some time to add a sequelize adapter! thank you!

**Pull request closed without merging**